source/MaterialXView/ShaderValidators/Glsl/GlslValidator.cpp

```diff
@@ -0,0 +1,1225 @@
+{
+// OpenGL Constants
+unsigned int GlslValidator::UNDEFINED_OPENGL_RESOURCE_ID = 0;
+int GlslValidator::UNDEFINED_OPENGL_PROGRAM_LOCATION = -1;
+int GlslValidator::ProgramInput::INVALID_OPENGL_TYPE = -1;
+
+// View information
+const float NEAR_PLANE = -100.0f;
+const float FAR_PLANE = 100.0f;
+
+GlslValidator::GlslValidator() :
+    _programId(0),
+    _colorTarget(0),
+    _depthTarget(0),
+    _frameBuffer(0),
+    _frameBufferWidth(256),
+    _frameBufferHeight(256),
+    _indexBuffer(0),
+    _indexBufferSize(0),
+    _vertexArray(0),
+    _dummyTexture(0),
+    _hwShader(nullptr),
+    _initialized(false)
+{
+    // Clear buffer ids to invalid identifier.
+    _attributeBufferIds.resize(ATTRIBUTE_COUNT);
+    std::fill(_attributeBufferIds.begin(), _attributeBufferIds.end(), UNDEFINED_OPENGL_RESOURCE_ID);
+}
+
+GlslValidator::~GlslValidator()
+{
+    deleteProgram();
+    deleteTarget();
+
+    GLUtilityContext* context = GLUtilityContext::get();
+    if (context)
+        GLUtilityContext::destroy();
+}
+
+void GlslValidator::setStage(const std::string& code, size_t stage)
+{
+    if (stage < HwShader::NUM_STAGES)
+    {
+        _stages[stage] = code;
+    }
+
+    // A stage change invalidates any cached parsed inputs
+    clearInputLists();
+    // For now a individual stage setting means to not use a HwShader.
+    // If / when injection of stages is allowed then this should not longer be true
+    _hwShader = nullptr;
+}
+
+
+void GlslValidator::setStages(const HwShaderPtr shader)
+{
+    if (!shader)
+    {
+        ShaderValidationErrorList errors;
+        throw ExceptionShaderValidationError("Cannot set stages using null hardware shader.", errors);
+    }
+
+    // Clear out any old data
+    clearStages();
+
+    // Extract out the shader code per stage
+    _hwShader = shader;
+    for (size_t i = 0; i < HwShader::NUM_STAGES; i++)
+    {
+        _stages[i] = _hwShader->getSourceCode(i);
+    }
+
+    // A stage change invalidates any cached parsed inputs
+    clearInputLists();
+}
+
+const std::string GlslValidator::getStage(size_t stage) const
+{
+    if (stage < HwShader::NUM_STAGES)
+    {
+        return _stages[stage];
+    }
+    return std::string();
+}
+
+void GlslValidator::clearStages()
+{
+    for (size_t i = 0; i < HwShader::NUM_STAGES; i++)
+    {
+        _stages[i].clear();
+    }
+
+    // Clearing stages invalidates any cached inputs
+    clearInputLists();
+}
+
+bool GlslValidator::haveValidStages() const
+{
+    // Need at least a pixel shader stage and a vertex shader stage
+    const std::string& vertexShaderSource = _stages[HwShader::VERTEX_STAGE];
+    const std::string& fragmentShaderSource = _stages[HwShader::PIXEL_STAGE];
+
+    return (vertexShaderSource.length() > 0 && fragmentShaderSource.length() > 0);
+}
+
+
+void GlslValidator::initialize()
+{
+    ShaderValidationErrorList errors;
+    const std::string errorType("OpenGL utilities initialization.");
+
+    if (!_initialized)
```

Comment:
This is one time initialization on top of which an arbitrary number of shaders can be compiled one and a time and validated. Uses GLEW and the custom utility code in this review just so we can start using OpenGL.